src/Workspaces/Core/MSBuild/MSBuild/MSBuildProjectLoader.Worker.cs

```diff
@@ -0,0 +1,492 @@
+                DiagnosticReportingOptions discoveredProjectOptions,
+                bool preferMetadataForReferencedProjects)
+            {
+                _workspace = workspace;
+                _diagnosticReporter = diagnosticReporter;
+                _pathResolver = pathResolver;
+                _projectFileLoaderRegistry = projectFileLoaderRegistry;
+                _buildManager = buildManager;
+                _baseDirectory = baseDirectory;
+                _requestedProjectPaths = requestedProjectPaths;
+                _globalProperties = globalProperties;
+                _projectMap = projectMap ?? ProjectMap.Create();
+                _progress = progress;
+                _requestedProjectOptions = requestedProjectOptions;
+                _discoveredProjectOptions = discoveredProjectOptions;
+                _preferMetadataForReferencedProjects = preferMetadataForReferencedProjects;
+                _projectIdToFileInfoMap = new Dictionary<ProjectId, ProjectFileInfo>();
+                _pathToDiscoveredProjectInfosMap = new Dictionary<string, ImmutableArray<ProjectInfo>>(PathUtilities.Comparer);
+                _projectIdToProjectReferencesMap = new Dictionary<ProjectId, List<ProjectReference>>();
+            }
+
+            private async Task<TResult> DoOperationAndReportProgressAsync<TResult>(ProjectLoadOperation operation, string projectPath, string targetFramework, Func<Task<TResult>> doFunc)
+            {
+                var watch = _progress != null
+                    ? Stopwatch.StartNew()
+                    : null;
+
+                var result = await doFunc().ConfigureAwait(false);
+
+                if (_progress != null)
+                {
+                    watch.Stop();
+                    _progress.Report(new ProjectLoadProgress(projectPath, operation, targetFramework, watch.Elapsed));
+                }
+
+                return result;
+            }
+
+            public async Task<ImmutableArray<ProjectInfo>> LoadAsync(CancellationToken cancellationToken)
+            {
+                var results = ImmutableArray.CreateBuilder<ProjectInfo>();
+                var processedPaths = new HashSet<string>(PathUtilities.Comparer);
+
+                _buildManager.Start();
+                try
+                {
+                    foreach (var projectPath in _requestedProjectPaths)
+                    {
+                        cancellationToken.ThrowIfCancellationRequested();
+
+                        if (!_pathResolver.TryGetAbsoluteProjectPath(projectPath, _baseDirectory, _requestedProjectOptions.OnPathFailure, out var absoluteProjectPath))
+                        {
+                            continue; // Failure should already be reported.
+                        }
+
+                        if (!processedPaths.Add(absoluteProjectPath))
+                        {
+                            // TODO: Report warning if there are duplicate project paths.
```

Comment:
Might as well do this, considering how much else is done in this PR :)

Reply:
Warning logged.